sw/airborne/firmwares/rotorcraft/guidance/guidance_indi.c

```diff
@@ -0,0 +1,228 @@
+{
+  roll_filt = roll_filt + roll_filtd / PERIODIC_FREQUENCY;
+  pitch_filt = pitch_filt + pitch_filtd / PERIODIC_FREQUENCY;
+
+  roll_filtd = roll_filtd + roll_filtdd / PERIODIC_FREQUENCY;
+  pitch_filtd = pitch_filtd + pitch_filtdd / PERIODIC_FREQUENCY;
+//   float cospsi = cosf(stateGetNedToBodyEulers_f()->psi);
+//   float sinpsi = sinf(stateGetNedToBodyEulers_f()->psi);
+
+  roll_filtdd = -roll_filtd * 2 * 0.55 * 50.0 + (stateGetNedToBodyEulers_f()->phi - roll_filt) * 50.0*50.0;
+  pitch_filtdd = -pitch_filtd * 2 * 0.55 * 50.0 + (stateGetNedToBodyEulers_f()->theta - pitch_filt) * 50.0*50.0;
+}
+
+void guidance_indi_calcG(struct FloatMat33 *Gmat) {
+
+  struct FloatEulers *euler = stateGetNedToBodyEulers_f();
+
+  float sphi = sinf(euler->phi);
+  float cphi = cosf(euler->phi);
+  float stheta = sinf(euler->theta);
+  float ctheta = cosf(euler->theta);
+  float spsi = sinf(euler->psi);
+  float cpsi = cosf(euler->psi);
+  float T = -9.81; //minus gravity is a good guesstimate of the thrust force
+//   float T = (filt_accelzn-9.81)/(cphi*ctheta); //calculate specific force in body z axis by using the accelerometer
+//   float T = filt_accelzbody; //if body acceleration is available, use that!
+
+  RMAT_ELMT(*Gmat, 0, 0) = (cphi*spsi - sphi*cpsi*stheta)*T;
+  RMAT_ELMT(*Gmat, 1, 0) = (-sphi*spsi*stheta - cpsi*cphi)*T;
+  RMAT_ELMT(*Gmat, 2, 0) = -ctheta*sphi*T;
+  RMAT_ELMT(*Gmat, 0, 1) = (cphi*cpsi*ctheta)*T;
+  RMAT_ELMT(*Gmat, 1, 1) = (cphi*spsi*ctheta)*T;
+  RMAT_ELMT(*Gmat, 2, 1) = -stheta*cphi*T;
+  RMAT_ELMT(*Gmat, 0, 2) = sphi*spsi + cphi*cpsi*stheta;
+  RMAT_ELMT(*Gmat, 1, 2) = cphi*spsi*stheta - cpsi*sphi;
+  RMAT_ELMT(*Gmat, 2, 2) = cphi*ctheta;
+}
+
+void stabilization_attitude_set_setpoint_rp_quat_f(bool_t in_flight)
+{
+  struct FloatQuat q_rp_cmd;
+  float_quat_of_eulers(&q_rp_cmd, &guidance_euler_cmd); //TODO this is a quaternion without yaw! add the desired yaw before you use it!
+
+  /* get current heading */
+  const struct FloatVect3 zaxis = {0., 0., 1.};
+  struct FloatQuat q_yaw;
+
+  float_quat_of_axis_angle(&q_yaw, &zaxis, stateGetNedToBodyEulers_f()->psi);
+
+  /* roll/pitch commands applied to to current heading */
+  struct FloatQuat q_rp_sp;
+  float_quat_comp(&q_rp_sp, &q_yaw, &q_rp_cmd);
+  float_quat_normalize(&q_rp_sp);
+
+  struct FloatQuat q_sp;
+
+  if (in_flight) {
+    /* get current heading setpoint */
+    struct FloatQuat q_yaw_sp;
+#if defined STABILIZATION_ATTITUDE_TYPE_INT
+    float_quat_of_axis_angle(&q_yaw_sp, &zaxis, ANGLE_FLOAT_OF_BFP(stab_att_sp_euler.psi));
+#else
+    float_quat_of_axis_angle(&q_yaw_sp, &zaxis, stab_att_sp_euler.psi);
+#endif
+
+    /* rotation between current yaw and yaw setpoint */
+    struct FloatQuat q_yaw_diff;
+    float_quat_comp_inv(&q_yaw_diff, &q_yaw_sp, &q_yaw);
+
+    /* compute final setpoint with yaw */
+    float_quat_comp_norm_shortest(&q_sp, &q_rp_sp, &q_yaw_diff);
+  } else {
+    QUAT_COPY(q_sp, q_rp_sp);
+  }
+
+  QUAT_BFP_OF_REAL(stab_att_sp_quat,q_sp);
```

Comment:
missing include `stabilization/stabilization_attitude_ref_quat_int.h`

and new line at end of file